bin/pre-install.js

```diff
@@ -0,0 +1,116 @@
+ *  * License, v. 2.0. If a copy of the MPL was not distributed with this
+ *   * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+ * This file is run when a user runs `yarn install`, before doing anything else.
+ * We check that various tools we need have the correct version. For now we
+ * check Node and Yarn. */
+
+// The "userAgent" is something like 'yarn/0.28.1 npm/? node/v8.4.0 linux x64'
+// when ran with yarn. With npm it's similar, without the `yarn` part of course.
+const userAgent = process.env.npm_config_user_agent;
+if (!userAgent) {
+  console.error('Error: this script cannot be run directly.');
+  process.exit(-1);
+}
+
+const agents = userAgent.split(' ').reduce((agents, agent) => {
+  const [key, value] = agent.split('/');
+  agents[key] = value;
+  return agents;
+}, {});
+
+const checks = [checkNode(), checkYarn()];
+
+const successful = checks.every(returnValue => returnValue);
+if (successful) {
+  console.log(
+    'All project requirements are satisfied, moving forward with the installation.'
+  );
+  process.exit(0);
+}
+process.exit(-1);
+
+// This function compares two string versions. This compares only major.minor
+// and disrespect any textual prevision (like pre or beta).
+function versionCompare(a, b) {
+  return a.localeCompare(b, undefined, { numeric: true });
+}
+
+function checkNode() {
+  // Node versions usually have a starting `v`.
+  const nodeVersion = agents.node.replace(/^v/, '');
+  const expectedNodeVersion = parseExpectedNodeVersion();
+  if (versionCompare(nodeVersion, expectedNodeVersion) < 0) {
+    console.error(
+      `This project expects at least Node version ${expectedNodeVersion}.`
+    );
+    console.error(
+      'You can use a tool like `nvm` to install and manage installed node versions.'
+    );
+    console.error(
+      'You can look at https://github.com/creationix/nvm to install this tool.\n'
+    );
+    return false;
+  }
+  return true;
+}
+
+function checkYarn() {
+  if (!('yarn' in agents)) {
+    console.error(
+      'This project uses Yarn instead of npm, please run `yarn install` instead of `npm install`.\n'
+    );
+    displayYarnVersionExplanation('Additionally ');
+    displayInstallationInformation();
+    return false;
+  }
+
+  if (agents.yarn) {
+    const version = agents.yarn;
+    // Poor man's check, but effective enough :)
+    if (version.startsWith('0.26.') || version.startsWith('0.27.')) {
+      displayYarnVersionExplanation();
+      displayInstallationInformation();
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function parseExpectedNodeVersion() {
+  // Let's fetch our minimal version from circleci's file
+  const fs = require('fs');
+  const circleConfig = fs.readFileSync('circle.yml', { encoding: 'utf8' });
+  const expectedNodeVersion = /version:\s+([\d.]+)/.exec(circleConfig)[1];
+  if (!expectedNodeVersion) {
+    throw new Error(`Couldn't extract the node version from circle.yml.`);
```

Comment:
Added this throw; this should show up in the CI (when it runs `yarn install`) and warn us when circle.yml is modified in a way that this script is broken.